File: tui/src/scala/tui/widgets/BarChartWidget.scala
package tui
package widgets

import tui.internal.ranges._
import tui.{Grapheme, Style}

/** Display multiple bars in a single widgets
  *
  * @param block
  *   Block to wrap the widget in
  * @param bar_width
  *   The width of each bar
  * @param bar_gap
  *   The gap between each bar
  * @param bar_set
  *   Set of symbols used to display the data
  * @param bar_style
  *   Style of the bars
  * @param value_style
  *   Style of the values printed at the bottom of each bar
  * @param label_style
  *   Style of the labels printed under each bar
  * @param style
  *   Style for the widget
  * @param data
  *   Slice of (label, value) pair to plot on the chart
  * @param max
  *   Value necessary for a bar to reach the maximum height (if no value is specified, the maximum value in the data is taken as reference)
  */
case class BarChartWidget(
    block: Option[BlockWidget] = None,
    bar_width: Int = 1,
    bar_gap: Int = 1,
    bar_set: symbols.bar.Set = symbols.bar.NINE_LEVELS,
    bar_style: Style = Style.DEFAULT,
    value_style: Style = Style.DEFAULT,
    label_style: Style = Style.DEFAULT,
    style: Style = Style.DEFAULT,
    data: Array[(String, Int)] = Array.empty,
    max: Option[Int] = None
) extends Widget {

  /** Values to display on the bar (computed when the data is passed to the widget)
    */
  private lazy val values: Array[Grapheme] = data.collect { case (_, v) => Grapheme(v.toString) }

  override def render(area: Rect, buf: Buffer): Unit = {
    buf.set_style(area, style)

    val chart_area: Rect = block match {
      case Some(b) =>
        val inner_area = b.inner(area)
        b.render(area, buf)
        inner_area
      case None => area
    }

    if (chart_area.height < 2) {
      return
    }

    val max = this.max.getOrElse(data.maxByOption { case (_, value) => value }.fold(0) { case (_, value) => value })

    val max_index = math.min(
      chart_area.width / (bar_width + bar_gap),
      data.length
    )

    case class Data(label: String, var value: Int)
    val data2 = this.data.take(max_index).map { case (l, v) => Data(l, v * (chart_area.height - 1) * 8 / math.max(max, 1)) }.zipWithIndex

    revRange(0, chart_area.height - 1) { j =>
      data2.foreach { case (d, i) =>
        val symbol = d.value match {
          case 0 => bar_set.empty
          case 1 => bar_set.one_eighth
          case 2 => bar_set.one_quarter
          case 3 => bar_set.three_eighths
          case 4 => bar_set.half
          case 5 => bar_set.five_eighths
          case 6 => bar_set.three_quarters
          case 7 => bar_set.seven_eighths
          case _ => bar_set.full
        }
        range(0, bar_width) { x =>
          buf
            .get(
              chart_area.left + i * (bar_width + bar_gap) + x,
              chart_area.top + j
            )
            .set_symbol(symbol)
            .set_style(bar_style)
          ()
        }

        if (d.value > 8) {
          d.value = d.value - 8
        } else {
          d.value = 0
        }
      }
    }
    data.take(max_index).zipWithIndex.foreach { case ((label, value), i) =>
      if (value != 0) {
        val value_label = values(i)
        val width = value_label.width
        if (width < bar_width) {
          buf.set_string(
            chart_area.left
              + i * (bar_width + bar_gap)
              + (bar_width - width) / 2,
            chart_area.bottom - 2,
            value_label.str,
            value_style
          )
        }
      }
      buf.set_stringn(
        chart_area.left + i * (bar_width + bar_gap),
        chart_area.bottom - 1,
        label,
        bar_width,
        label_style
      );

    }
  }
}

File: tui/src/java/tui/widgets/BarChartWidget.java
package tui.widgets;

import java.util.Optional;
import tui.Buffer;
import tui.Grapheme;
import tui.Rect;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;

/// Display multiple bars in a single widgets
public final class BarChartWidget implements Widget {
  public final Optional<BlockWidget> block;
  public final int barWidth;
  public final int barGap;
  public final Symbols.bar.Set barSet;
  public final Style barStyle;
  public final Style valueStyle;
  public final Style labelStyle;
  public final Style style;
  public final LabelValue[] data;
  public final Optional<Integer> max;

  public BarChartWidget(
      Optional<BlockWidget> block,
      int barWidth,
      int barGap,
      Symbols.bar.Set barSet,
      Style barStyle,
      Style valueStyle,
      Style labelStyle,
      Style style,
      LabelValue[] data,
      Optional<Integer> max) {
    this.block = block;
    this.barWidth = barWidth;
    this.barGap = barGap;
    this.barSet = barSet;
    this.barStyle = barStyle;
    this.valueStyle = valueStyle;
    this.labelStyle = labelStyle;
    this.style = style;
    this.data = data;
    this.max = max;
  }

  public static BarChartWidget empty() {
    return new BarChartWidget(
        Optional.empty(),
        1,
        1,
        Symbols.bar.NINE_LEVELS,
        Style.DEFAULT,
        Style.DEFAULT,
        Style.DEFAULT,
        Style.DEFAULT,
        new LabelValue[0],
        Optional.empty());
  }

  public BarChartWidget withBlock(BlockWidget b) {
    return new BarChartWidget(
        Optional.of(b), barWidth, barGap, barSet, barStyle, valueStyle, labelStyle, style, data, max);
  }

  public BarChartWidget withBarWidth(int w) {
    return new BarChartWidget(
        block, w, barGap, barSet, barStyle, valueStyle, labelStyle, style, data, max);
  }

  public BarChartWidget withBarGap(int g) {
    return new BarChartWidget(
        block, barWidth, g, barSet, barStyle, valueStyle, labelStyle, style, data, max);
  }

  public BarChartWidget withBarSet(Symbols.bar.Set s) {
    return new BarChartWidget(
        block, barWidth, barGap, s, barStyle, valueStyle, labelStyle, style, data, max);
  }

  public BarChartWidget withBarStyle(Style s) {
    return new BarChartWidget(
        block, barWidth, barGap, barSet, s, valueStyle, labelStyle, style, data, max);
  }

  public BarChartWidget withValueStyle(Style s) {
    return new BarChartWidget(
        block, barWidth, barGap, barSet, barStyle, s, labelStyle, style, data, max);
  }

  public BarChartWidget withLabelStyle(Style s) {
    return new BarChartWidget(
        block, barWidth, barGap, barSet, barStyle, valueStyle, s, style, data, max);
  }

  public BarChartWidget withStyle(Style s) {
    return new BarChartWidget(
        block, barWidth, barGap, barSet, barStyle, valueStyle, labelStyle, s, data, max);
  }

  public BarChartWidget withData(LabelValue[] d) {
    return new BarChartWidget(
        block, barWidth, barGap, barSet, barStyle, valueStyle, labelStyle, style, d, max);
  }

  public BarChartWidget withMax(int m) {
    return new BarChartWidget(
        block, barWidth, barGap, barSet, barStyle, valueStyle, labelStyle, style, data, Optional.of(m));
  }

  @Override
  public void render(Rect area, Buffer buf) {
    buf.setStyle(area, style);

    Rect chartArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      chartArea = innerArea;
    } else {
      chartArea = area;
    }

    if (chartArea.height() < 2) {
      return;
    }
    // v0.24.0 #525: avoid divide by zero on zero-width bars or empty data.
    if (chartArea.area() == 0 || data.length == 0 || barWidth == 0) {
      return;
    }

    int maxValue;
    if (max.isPresent()) {
      maxValue = max.get();
    } else {
      int m = 0;
      for (LabelValue lv : data) {
        if (lv.value() > m) m = lv.value();
      }
      maxValue = m;
    }

    int maxIndex = Math.min(chartArea.width() / (barWidth + barGap), data.length);

    // Mutable per-index decreasing values, with original labels.
    int[] data2 = new int[maxIndex];
    for (int i = 0; i < maxIndex; i++) {
      data2[i] = data[i].value() * (chartArea.height() - 1) * 8 / Math.max(maxValue, 1);
    }

    Ranges.revRange(
        0,
        chartArea.height() - 1,
        j -> {
          for (int i = 0; i < data2.length; i++) {
            int v = data2[i];
            String symbol =
                switch (v) {
                  case 0 -> barSet.empty();
                  case 1 -> barSet.oneEighth();
                  case 2 -> barSet.oneQuarter();
                  case 3 -> barSet.threeEighths();
                  case 4 -> barSet.half();
                  case 5 -> barSet.fiveEighths();
                  case 6 -> barSet.threeQuarters();
                  case 7 -> barSet.sevenEighths();
                  default -> barSet.full();
                };
            int idx = i;
            Ranges.range(
                0,
                barWidth,
                x -> buf.get(
                        chartArea.left() + idx * (barWidth + barGap) + x,
                        chartArea.top() + j)
                    .setSymbol(symbol)
                    .setStyle(barStyle));
            if (v > 8) {
              data2[i] = v - 8;
            } else {
              data2[i] = 0;
            }
          }
        });

    // Values to display on the bar (computed when the data is passed to the widget)
    Grapheme[] values = new Grapheme[maxIndex];
    for (int i = 0; i < maxIndex; i++) {
      values[i] = new Grapheme(Integer.toString(data[i].value()));
    }
    for (int i = 0; i < maxIndex; i++) {
      LabelValue lv = data[i];
      String label = lv.label();
      int value = lv.value();
      if (value != 0) {
        Grapheme valueLabel = values[i];
        int width = valueLabel.width();
        if (width < barWidth) {
          buf.setString(
              chartArea.left() + i * (barWidth + barGap) + (barWidth - width) / 2,
              chartArea.bottom() - 2,
              valueLabel.str,
              valueStyle);
        }
      }
      buf.setStringn(
          chartArea.left() + i * (barWidth + barGap),
          chartArea.bottom() - 1,
          label,
          barWidth,
          labelStyle);
    }
  }

  public record LabelValue(String label, int value) {}
}
